### PR TITLE
misc: fix compilation with Solaris Studio

### PR DESCRIPTION
**Pull request by hzhou**
## Pull Request Description

The current Solaris studio nightly tests fail with 
```
"src/util/mpir_hwtopo.c", line 246: internal compiler error: __builtin_expect undefined
cc: acomp failed for src/util/mpir_hwtopo.c
make[2]: *** [src/util/lib_libmpi_la-mpir_hwtopo.lo] Error 1
make[2]: *** Waiting for unfinished jobs....
```
Additional failures include not able to link `mpichversion` and `libmpifort.so` due to including `mpiimpl.h`, which contains static inline functions that references internal functions in `libmpi.so`.
[skip warnings]
## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [ ] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

### CODE REVIEW

**Comment by hzhou** [comment]
test:mpich/custom
netmod: ch4:ofi
compiler: solstudio

**Comment by hzhou** [comment]
test:mpich/custom
netmod: ch4:ofi
compiler: solstudio

File: src/util/mpir_hwtopo.c [hunk 1]
```diff
@@ -52,49 +52,45 @@ typedef enum {
...
+    return depth;
+}
+
+static int HWTOPO_GET_INDEX(MPIR_hwtopo_gid_t gid)
+{
+    int idx;
+    idx = (gid & HWTOPO_GID_INDEX_MASK);
+    idx = (idx >> HWTOPO_GID_INDEX_SHIFT);
+    return idx;
+}
```
Comment (raffenet) [inline src/util/mpir_hwtopo.c:93]: 👍. I need this for 3.4.x, too.

**Comment by hzhou** [comment]
Clueless, so try it again 😞 :
test:mpich/custom
netmod: ch4:ofi
compiler: solstudio

**Comment by hzhou** [comment]
test:mpich/custom
netmod: ch4:ofi
compiler: solstudio

**Comment by hzhou** [comment]
test:mpich/custom
netmod: ch4:ofi
compiler: solstudio

**Comment by hzhou** [comment]
test:mpich/custom
netmod: ch4:ofi
compiler: solstudio

**Comment by hzhou** [comment]
test:mpich/ch4/most
test:mpich/ch3/most

**Comment by hzhou** [comment]
test:mpich/ch4/most
test:mpich/ch3/most

File: src/binding/fortran/mpif_h/mpi_fortimpl.h [hunk 2]
```diff
@@ -7,6 +7,10 @@
 #define MPI_FORTIMPL_H_INCLUDED
 
 #include "mpichconf.h"
+#include "mpi.h"
```
Comment (raffenet) [inline src/binding/fortran/mpif_h/mpi_fortimpl.h:10]: I think this is probably good. Looks like we need to `#include <sys/types.h>` somewhere to get `ssize_t` on BSD and macOS.

Comment (hzhou) [reply src/binding/fortran/mpif_h/mpi_fortimpl.h:10]: Added.

**Comment by hzhou** [comment]
test:mpich/ch3/most

**Review by raffenet — APPROVED** [review]
Looks good as long as tests pass.

**Pull request merged**